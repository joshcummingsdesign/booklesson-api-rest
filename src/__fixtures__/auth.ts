import { IJwtPayload } from '../auth/interfaces';
import { Auth } from '../auth/entities';

export const auth: Omit<Auth, 'user'> = {
  userId: 1,
  password: 'Changeme!',
};

export const jwtPayload: IJwtPayload = {
  access_token:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6Impvc2hjdW1taW5nc2Rlc2lnbkBnbWFpbC5jb20iLCJzdWIiOjEsImlhdCI6MTU5MTc0ODUyMywiZXhwIjoxNTkxNzQ4NTMzfQ.jgWsQFZOdYX0e3oPzAByc3uwDPhYNtfNsk1a8Akp5uE',
};
