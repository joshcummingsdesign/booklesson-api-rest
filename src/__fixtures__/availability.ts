import { Availability } from '../availability/entities';
import { AddAvailabilityDto } from '../availability/dto';

export const availability: Partial<Availability> = {
  id: 999,
  userId: 999,
  datetime: '2020-06-21T00:00:00.000Z',
  available: true,
};

export const availabilities: Partial<Availability>[] = [
  {
    id: 998,
    userId: 999,
    datetime: '2020-06-20T23:00:00.000Z',
    available: false,
  },
  {
    id: 999,
    userId: 999,
    datetime: '2020-06-21T00:00:00.000Z',
    available: true,
  },
];

export const addAvailabilityDto: AddAvailabilityDto = {
  datetime: '2020-06-11T10:00:00Z',
};
