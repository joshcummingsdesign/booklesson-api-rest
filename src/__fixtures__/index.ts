export * from './auth';
export * from './user';
export * from './role';
export * from './instrument';
export * from './lesson';
export * from './availability';
