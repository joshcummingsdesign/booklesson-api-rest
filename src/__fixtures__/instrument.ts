import { Instrument } from '../instrument/entities';

export const instrument: Partial<Instrument> = {
  id: 996,
  name: 'guitar',
};

export const instruments: Partial<Instrument>[] = [
  {
    id: 996,
    name: 'guitar',
  },
  {
    id: 997,
    name: 'piano',
  },
  {
    id: 998,
    name: 'bass',
  },
  {
    id: 999,
    name: 'drums',
  },
];
