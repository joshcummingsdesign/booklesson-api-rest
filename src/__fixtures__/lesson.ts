import { Lesson } from '../lesson/entities';
import { CreateLessonDto } from '../lesson/dto';

type PartialLesson = Omit<Lesson, 'student' | 'teacher'>;

export const lesson: PartialLesson = {
  id: 1,
  studentId: 1,
  teacherId: 2,
  date: '2020-06-10',
  time: '09:00:00',
};

export const lessons: PartialLesson[] = [lesson];

export const createLessonDto: CreateLessonDto = {
  teacherId: 2,
  date: '2020-06-10',
  time: '09:00:00',
};
