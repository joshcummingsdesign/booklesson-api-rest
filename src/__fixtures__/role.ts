import { Role } from '../role/entities';

export const studentRole: Role = {
  id: 998,
  name: 'student',
} as Role;

export const teacherRole: Role = {
  id: 999,
  name: 'teacher',
} as Role;
