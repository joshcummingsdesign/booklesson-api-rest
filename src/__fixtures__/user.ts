import { User } from '../user/entities';
import { CreateUserDto, UpdateUserDto } from '../user/dto';
import { auth } from './auth';
import { studentRole, teacherRole } from './role';
import { instrument } from './instrument';
import { Role } from '../role/entities';
import { Instrument } from '../instrument/entities';

export const student: Partial<User> = {
  id: 998,
  firstName: 'Josh',
  lastName: 'Cummings',
  email: 'joshcummingsdesign@gmail.com',
  roleId: 998,
  instrumentId: 996,
  role: studentRole as Role,
  instrument: instrument as Instrument,
};

export const teacher: Partial<User> = {
  id: 999,
  firstName: 'Kurt',
  lastName: 'Rosenwinkel',
  email: 'kurtrosenwinkel@gmail.com',
  roleId: 999,
  instrumentId: 996,
  role: teacherRole as Role,
  instrument: instrument as Instrument,
};

export const studentAuth = { ...student, auth } as User;

export const teacherAuth = {
  ...teacher,
  auth: { ...auth, userId: teacher.id },
} as User;

export const users: Partial<User>[] = [student, teacher];

export const students: Partial<User>[] = [student];

export const teachers: Partial<User>[] = [teacher];

export const createUserDto: CreateUserDto = {
  firstName: 'Josh',
  lastName: 'Cummings',
  email: 'joshcummingsdesign@gmail.com',
  password: 'Changeme1!',
  instrument: 'guitar',
  role: 'student',
};

export const updateUserDto: UpdateUserDto = {
  firstName: 'Josh',
  lastName: 'Cummings',
  instrument: 'guitar',
};

export const createNewUser: CreateUserDto = {
  firstName: 'Victor',
  lastName: 'Wooten',
  email: 'victorwooten@gmail.com',
  password: 'Changeme1!',
  instrument: 'bass',
  role: 'teacher',
};

export const updateNewUser: UpdateUserDto = {
  firstName: 'Victor',
  lastName: 'Wooten',
  instrument: 'bass',
};
