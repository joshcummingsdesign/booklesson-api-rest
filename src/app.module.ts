import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { TypeOrmModule } from '@nestjs/typeorm';
import { validationSchema } from './config';
import { CryptoModule } from './crypto/crypto.module';
import { AuthModule } from './auth/auth.module';
import { RoleModule } from './role/role.module';
import { InstrumentModule } from './instrument/instrument.module';
import { UserModule } from './user/user.module';
import { LessonModule } from './lesson/lesson.module';
import { AvailabilityModule } from './availability/availability.module';

@Module({
  imports: [
    ConfigModule.forRoot({ isGlobal: true, validationSchema }),
    TypeOrmModule.forRoot(),
    CryptoModule,
    AuthModule,
    RoleModule,
    InstrumentModule,
    UserModule,
    LessonModule,
    AvailabilityModule,
  ],
})
export class AppModule {}
