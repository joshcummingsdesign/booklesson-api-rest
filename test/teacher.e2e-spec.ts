import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';
import { teacherAuth, addAvailabilityDto } from '../src/__fixtures__';

describe('Teacher (e2e)', () => {
  let app: INestApplication;
  let token: string;
  let realDateNow: () => number;

  beforeAll(async () => {
    realDateNow = Date.now.bind(global.Date);
    const dateNowStub = jest.fn(() => 1592164576240);
    global.Date.now = dateNowStub;

    const module: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = module.createNestApplication();

    app.init();

    token = await request(app.getHttpServer())
      .post('/api/v1/auth/login')
      .send({
        username: teacherAuth.email,
        password: teacherAuth.auth.password,
      })
      .then((res) => `Bearer ${res.body.access_token}`);
  });

  afterAll(() => {
    global.Date.now = realDateNow;
  });

  it('should view all lessons', async () => {
    const res = await request(app.getHttpServer())
      .get('/api/v1/lessons/self')
      .set('Authorization', token)
      .expect(200);

    expect(res.body).toMatchSnapshot();
  });

  it('should create, read, delete availability', async () => {
    const createRes = await request(app.getHttpServer())
      .post('/api/v1/availability')
      .send(addAvailabilityDto)
      .set('Authorization', token)
      .expect(201);

    expect(createRes.body).toMatchSnapshot();

    const readRes = await request(app.getHttpServer())
      .get('/api/v1/availability/self')
      .set('Authorization', token)
      .expect(200);

    expect(readRes.body).toMatchSnapshot();

    const deleteRes = await request(app.getHttpServer())
      .delete(`/api/v1/availability/self/${createRes.body.id}`)
      .set('Authorization', token)
      .expect(200);

    expect(deleteRes.body).toMatchSnapshot();
  });

  afterAll(async () => {
    await app.close();
  });
});
