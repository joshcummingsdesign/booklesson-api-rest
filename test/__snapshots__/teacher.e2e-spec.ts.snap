// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teacher (e2e) should create, read, delete availability 1`] = `
Object {
  "available": true,
  "datetime": "2020-06-20T11:00:00-07:00",
  "id": 1,
  "userId": 999,
}
`;

exports[`Teacher (e2e) should create, read, delete availability 2`] = `
Array [
  Object {
    "available": false,
    "datetime": "2020-06-20T09:00:00-07:00",
    "id": 998,
    "userId": 999,
  },
  Object {
    "available": true,
    "datetime": "2020-06-20T10:00:00-07:00",
    "id": 999,
    "userId": 999,
  },
  Object {
    "available": true,
    "datetime": "2020-06-20T11:00:00-07:00",
    "id": 1,
    "userId": 999,
  },
]
`;

exports[`Teacher (e2e) should create, read, delete availability 3`] = `
Object {
  "available": true,
  "datetime": "2020-06-20T11:00:00-07:00",
  "id": 1,
  "userId": 999,
}
`;

exports[`Teacher (e2e) should view all lessons 1`] = `
Array [
  Object {
    "datetime": "2020-06-20T09:00:00-07:00",
    "id": 999,
    "studentId": 998,
    "teacherId": 999,
  },
]
`;
